crates/subspace-farmer/src/ws_rpc_server.rs

```diff
@@ -1,19 +1,26 @@
 use crate::object_mappings::ObjectMappings;
-use crate::single_disk_farm::SingleDiskFarmPieceGetter;
 use async_trait::async_trait;
 use hex_buffer_serde::{Hex, HexForm};
 use jsonrpsee::core::error::Error;
 use jsonrpsee::proc_macros::rpc;
 use parity_scale_codec::{Compact, CompactLen, Decode, Encode};
 use serde::{Deserialize, Serialize};
 use std::ops::{Deref, DerefMut};
+use std::sync::Arc;
 use subspace_archiving::archiver::{Segment, SegmentItem};
 use subspace_core_primitives::{Piece, PieceIndex, PieceIndexHash, Sha256Hash};
 use tracing::{debug, error};
 
 /// Maximum expected size of one object in bytes
 const MAX_OBJECT_SIZE: usize = 5 * 1024 * 1024;
 
+/// Something that can be used to get decoded pieces by index
+pub trait PieceGetter {
+    /// Get piece
+    fn get_piece(&self, piece_index: PieceIndex, piece_index_hash: PieceIndexHash)
+        -> Option<Piece>;
```

Comment:
Do we need to ask for `piece_index_hash` here also? I think we can derive it from `piece_index` anyway

Reply:
We can, but then we need to hash piece as many times as we have plots (in the worst case), it felt like meaningful micro-optimization to hash it once outside and send down in already hashed form. API isn't as nice because of that, I agree.